File: src/bilisama/ui/audio.py
"""Who holds the microphone and the speaker, and the socket they travel on.

The point of moving audio into the page is echo cancellation. Chromium has it
(`getUserMedia({echoCancellation: true})`), the shell IS Chromium, and without
it the streamer has to wear headphones or give up barge-in during playback —
the trade dev_talk has been apologising for on every start.

There is a hard precondition, and it is what shapes this module: the canceller
can only subtract audio the browser itself rendered. Capturing in the page
while playback stays in sounddevice would leave it with no reference signal
and no effect at all. So both ends move, together, or neither does.

Two decisions worth stating because their alternatives look reasonable:

Audio gets its own socket. The control hub broadcasts to every client and
drops the oldest frame when a queue fills — right for sticky state, wrong for
samples, and 48 KB/s of PCM through it would starve the panel. This one is
point to point with the owner and carries binary frames.

Exactly one owner at a time, and the shell outranks a browser tab. Two
capture streams would fight over the device; two playback queues would talk
over each other. The shell is the product path and a tab is a developer's
convenience, so when both are open the shell wins and the tab is told plainly
that it is watching rather than left looking broken.
"""

from __future__ import annotations

from array import array
from collections import deque
from collections.abc import Callable
from dataclasses import dataclass
from typing import Literal, Protocol

from bilisama.obs.logging import get_logger

__all__ = [
    "AudioBroker",
    "AudioOwner",
    "EchoProbe",
    "EchoReading",
    "LocalAudio",
    "PlaybackTally",
]

log = get_logger(__name__)

# Ranked, best first. A claim only displaces a strictly weaker holder.
AudioOwner = Literal["shell", "browser"]
_RANK: dict[AudioOwner, int] = {"shell": 2, "browser": 1}


class LocalAudio(Protocol):
    """The sounddevice pair the broker parks while a page holds the devices.

    dev-talk implements this over its own microphone pump and speaker. Both
    calls run off the loop: releasing a PortAudio stream blocks, and doing it
    here would freeze the very loop that has to keep the socket alive.
    """

    async def suspend(self) -> None: ...

    async def resume(self) -> None: ...


class AudioBroker:
    """At most one holder of the microphone and speaker.

    Nobody holding is the normal state, not a failure: `--no-ui`, a shell that
    was never installed, a page nobody opened. Local audio simply keeps
    running, which is what every session did before this existed.
    """

    __slots__ = ("_announce", "_close", "_flush", "_local", "_owner", "_send")

    def __init__(
        self,
        *,
        local: LocalAudio | None = None,
        announce: Callable[[AudioOwner | None], None] | None = None,
    ) -> None:
        """Args:
        local: The sounddevice pair to park and restore. None in tests, and
            in any run that never had one.
        announce: Told who holds the devices, on every change. The panel shows
            it; a tab that is only watching needs to know why it has no sound.
        """
        self._local = local
        self._announce = announce
        self._owner: AudioOwner | None = None
        self._send: Callable[[bytes], None] | None = None
        self._close: Callable[[], None] | None = None
        self._flush: Callable[[], None] | None = None

    @property
    def owner(self) -> AudioOwner | None:
        return self._owner

    @property
    def local_is_live(self) -> bool:
        """True when the sounddevice pair is the one being heard."""
        return self._owner is None

    async def claim(
        self,
        who: AudioOwner,
        *,
        send: Callable[[bytes], None],
        close: Callable[[], None] | None = None,
        flush: Callable[[], None] | None = None,
    ) -> bool:
        """Hand the devices to a page, if it outranks whoever has them.

        Args:
            who: shell or browser; a shell displaces a tab, never the reverse.
            send: How to push downlink PCM to this client.
            close: How to hang up on this client when something stronger takes
                the devices. Without it a displaced tab keeps a live capture
                running on a socket that will never be fed again, and its panel
                goes on claiming to hold devices it lost.
            flush: How to drop everything queued for this client and tell it to
                stop playing. See flush().

        Returns:
            True if the caller now owns the devices. False means someone
            stronger has them and this client should watch, not listen.
        """
        if self._owner is not None and _RANK[who] <= _RANK[self._owner]:
            log.info("audio.claim_refused", who=who, holder=self._owner)
            return False
        first = self._owner is None
        displaced = self._close if not first else None
        self._owner = who
        self._send = send
        self._close = close
        self._flush = flush
        if displaced is not None:
            log.info("audio.displaced", by=who)
            displaced()
        if first and self._local is not None:
            # Only on the way in from nobody: a shell taking over from a tab
            # must not restart devices that are already parked.
            await self._local.suspend()
        log.info("audio.claimed", who=who)
        self._notify()
        return True

    async def release(self, who: AudioOwner) -> None:
        """Give the devices back. Ignored from a client that never had them."""
        if self._owner != who:
            return
        self._owner = None
        self._send = None
        self._close = None
        self._flush = None
        if self._local is not None:
            await self._local.resume()
        log.info("audio.released", who=who)
        self._notify()

    def flush(self) -> None:
        """Barge-in: drop what is queued and tell the page to stop.

        The control socket already carries playback.clear, and on its own that
        is not enough — it is a DIFFERENT connection, so audio the server had
        already queued arrives after the page has cleared and gets scheduled
        all over again. What the streamer hears is the text stopping while the
        voice carries on for a beat, which is exactly what a barge-in must not
        feel like.

        So the stop travels with the audio: everything queued goes, and the
        marker behind it is ordered after every sample already in flight.
        """
        if self._flush is not None:
            self._flush()

    def play(self, pcm: bytes) -> None:
        """Send one downlink chunk to the owner, or drop it.

        Dropping is right during the gap between a tab closing and the local
        pair coming back: those samples have nowhere to go, and holding them
        would only play them late.
        """
        if self._send is not None:
            self._send(pcm)

    def announce_through(self, announce: Callable[[AudioOwner | None], None]) -> None:
        """Say who holds the devices this way from now on.

        Set by create_ui_app rather than by whoever built the broker: the hub
        it broadcasts on is the app's, and a caller that wires one but not the
        other leaves every panel waiting on an answer that never comes.
        """
        self._announce = announce

    def _notify(self) -> None:
        if self._announce is not None:
            self._announce(self._owner)


class PlaybackTally:
    """How many segments the page still has to play, and the gate that reads it.

    Backlog item 41 exists for this class. The receipts are per SEGMENT: one
    reply is scheduled as many buffers, and between any two of them there is an
    instant where the previous has ended and the next has not started. Treating
    that instant as "finished speaking" opens the floor mid-sentence and lets
    the scheduler dispatch over her — which is precisely how the 106 seconds of
    unplayed audio accumulated the first time, just arriving from a different
    direction.

    So the gate follows a COUNT, not the last event. It closes on the first
    segment and opens only when the last one is done.
    """

    __slots__ = ("_notify", "_on_playback", "_outstanding")

    def __init__(
        self,
        *,
        on_playback: Callable[[bool], None],
        notify: Callable[[], None],
    ) -> None:
        """Args:
        on_playback: SpeakingFloor.on_playback — the gate itself, unchanged
            product code that only wanted a producer.
        notify: Scheduler.notify. State gates release on events, and playback
            finishing is not one the link ever sends.
        """
        self._on_playback = on_playback
        self._notify = notify
        self._outstanding = 0

    @property
    def outstanding(self) -> int:
        return self._outstanding

    def started(self) -> None:
        self._outstanding += 1
        if self._outstanding == 1:
            self._on_playback(True)

    def ended(self) -> None:
        # Clamped: a receipt for a segment that was already cancelled would
        # otherwise drive the count negative and wedge the gate shut.
        self._outstanding = max(0, self._outstanding - 1)
        if self._outstanding == 0:
            self._on_playback(False)
            self._notify()

    def cancelled(self) -> None:
        """A barge-in took everything scheduled. Nothing is outstanding now."""
        if self._outstanding == 0:
            return
        self._outstanding = 0
        self._on_playback(False)
        self._notify()


# One envelope point per 20 ms. The trick that makes this cheap: buckets are a
# TIME grid, so a 24 kHz downlink and a 16 kHz uplink both land on the same
# 50 Hz axis and no resampling is needed to compare them.
_BUCKET_MS = 20
# Six seconds of history, and a lag search covering two. The page schedules
# playback ahead of itself, so what we sent is heard some unknown amount later
# — anywhere inside its buffer. Searching wide costs 300x100 multiply-adds once
# a second, which is nothing.
_WINDOW = 6000 // _BUCKET_MS
_MAX_LAG = 2000 // _BUCKET_MS
# Below this the window is mostly silence and a correlation off it means
# nothing. Peak amplitude out of 32767.
_QUIET = 300


@dataclass(frozen=True, slots=True)
class EchoReading:
    """How much of her own voice came back through the microphone."""

    leak: float
    """0..1. Correlation between what we played and what the mic heard."""
    lag_ms: int
    """Roughly where the match sat. A hint, not a measurement: speech segments
    run long against a 20 ms grid, so a shifted envelope still overlaps itself
    heavily and neighbouring lags score almost the same. Useful for telling a
    direct acoustic path from one routed through another application, and not
    for anything finer."""
    verdict: Literal["ok", "suspect", "leaking"]


class EchoProbe:
    """Is she hearing herself? The one question Chromium cannot answer.

    Echo cancellation only subtracts audio the browser itself rendered
    (media_switches.cc:506-509). When her voice reaches the microphone by any
    other route — OBS monitoring it back out to the speakers is the one this
    product actively recommends — the canceller is working perfectly and the
    echo is there anyway. Nothing errors. The panel says cancellation is on.
    The only symptom is turn detection firing on her own voice, which reads as
    a broken endpointer and sends people tuning a VAD threshold that was never
    the problem (plan section 11).

    So this watches both streams and asks whether the microphone's loudness
    envelope contains a delayed copy of the reply's. It never cancels
    anything — it only tells you whether something else should have.

    An indicator, not proof. The streamer answering her produces some genuine
    correlation, so the number is reported rather than turned into a verdict
    the caller cannot see behind.

    Amplitude-blind by construction: the correlation is normalised, so a faint
    echo and a loud one both read near 1. That is the right choice here —
    "any of her voice is coming back" is the question, and a quiet leak
    false-triggers turn detection just as well as a loud one.

    One honest limitation, found staging a leak against the real endpoint: the
    fault this looks for eats its own evidence. A real echo trips the server's
    turn detection, the reply is cancelled, and her speech stops — so the
    window fills with silence and the correlation drops. It reads clearly over
    a stream that keeps being interrupted, and reads weakly over the single
    interruption that started it.
    """

    def status(self) -> dict[str, object]:
        """A health-registry card. Shows up in the panel alongside the others."""
        reading = self.reading()
        if reading is None:
            return {"ok": True, "state": "她还没开口，没什么可判断的"}
        wording = {
            "ok": "没听见她自己",
            "suspect": "疑似听见了她自己",
            "leaking": "她的声音正漏进麦克风",
        }[reading.verdict]
        return {
            # suspect stays healthy: the streamer answering her is genuine
            # correlation, and a card that cries wolf gets ignored.
            "ok": reading.verdict != "leaking",
            "state": wording,
            "leak": reading.leak,
            "lag_ms": reading.lag_ms,
        }

    __slots__ = ("_captured", "_played")

    def __init__(self) -> None:
        self._played = _Envelope(24000)
        self._captured = _Envelope(16000)

    def note_played(self, pcm: bytes) -> None:
        """One downlink chunk, on its way to whoever holds the speaker."""
        self._played.feed(pcm)

    def note_captured(self, pcm: bytes) -> None:
        """One uplink chunk, as the microphone heard it."""
        self._captured.feed(pcm)

    def reading(self) -> EchoReading | None:
        """The current estimate, or None when there is nothing to judge.

        None means she has not been speaking, so the microphone has no echo of
        hers to contain — silence is not evidence of anything.
        """
        played = self._played.window(_WINDOW)
        heard = self._captured.window(_WINDOW)
        span = min(len(played), len(heard))
        if span < _MAX_LAG + _WINDOW // 4:
            return None
        played, heard = played[-span:], heard[-span:]
        if max(played) < _QUIET:
            return None

        best, best_lag = 0.0, 0
        for lag in range(_MAX_LAG):
            # The reply leads: what we sent at t is heard at t + lag.
            score = _correlate(heard[lag:], played[: span - lag])
            if score > best:
                best, best_lag = score, lag
        verdict: Literal["ok", "suspect", "leaking"] = (
            "leaking" if best >= 0.5 else "suspect" if best >= 0.25 else "ok"
        )
        return EchoReading(leak=round(best, 3), lag_ms=best_lag * _BUCKET_MS, verdict=verdict)


class _Envelope:
    """Peak amplitude per 20 ms bucket, built as chunks arrive."""

    __slots__ = ("_buckets", "_carry", "_per_bucket", "_step")

    def __init__(self, rate: int) -> None:
        self._per_bucket = rate * _BUCKET_MS // 1000
        # Every eighth sample. A peak envelope survives decimation fine, and
        # this runs on the audio path — 2000 abs() a second beats 20000.
        self._step = 8
        self._carry = b""
        self._buckets: deque[int] = deque(maxlen=_WINDOW)

    def feed(self, pcm: bytes) -> None:
        data = self._carry + pcm
        width = self._per_bucket * 2
        whole = len(data) // width
        for i in range(whole):
            block = data[i * width : (i + 1) * width]
            samples = array("h")
            samples.frombytes(block)
            peak = 0
            for j in range(0, len(samples), self._step):
                value = samples[j]
                # -32768 has no positive counterpart; clamp rather than let
                # abs() hand back a number that does not fit the range.
                peak = max(peak, 32767 if value == -32768 else abs(value))
            self._buckets.append(peak)
        self._carry = data[whole * width :]

    def window(self, size: int) -> list[int]:
        return list(self._buckets)[-size:]


def _correlate(left: list[int], right: list[int]) -> float:
    """Normalised correlation of two envelopes, 0 when either is flat."""
    span = min(len(left), len(right))
    if span < 8:
        return 0.0
    left, right = left[:span], right[:span]
    mean_l = sum(left) / span
    mean_r = sum(right) / span
    cov = var_l = var_r = 0.0
    for a, b in zip(left, right, strict=True):
        da, db = a - mean_l, b - mean_r
        cov += da * db
        var_l += da * da
        var_r += db * db
    if var_l <= 0 or var_r <= 0:
        return 0.0
    return max(0.0, float(cov / (var_l * var_r) ** 0.5))

File: src/bilisama/ui/web/js/audio.js
// The microphone and the speaker, on the page where the echo canceller lives.
//
// Why here and not in Python: Chromium's canceller can only subtract audio the
// browser itself rendered. Capturing here while playback stayed in sounddevice
// would give it no reference signal and no effect, and the streamer would keep
// wearing headphones. So both ends live in this file.
//
// Two clocks, on purpose. Capture runs at whatever the device gave us and the
// worklet resamples down to 16k. Playback schedules 24k buffers against the
// context's own clock, because `nextStartTime` bookkeeping is the only way to
// get gapless streaming out of Web Audio — a source started "now" for each
// chunk leaves an audible seam every time the main thread is busy.

const DOWNLINK_RATE = 24000; // what the providers send (plan section 3.1)

export function createAudio({ onOwner }) {
  // Receipts go out on THIS socket, not the control one. They describe the
  // audio that just arrived here, and the control socket is a separate
  // connection opening on its own schedule — send a playback.started before it
  // is up and the frame is dropped, leaving the floor gate believing she is
  // silent while she talks. Measured: on a second page load the audio socket
  // consistently won the race and every started receipt vanished.
  const report = (event, data = {}) => {
    if (socket && socket.readyState === WebSocket.OPEN) {
      socket.send(JSON.stringify({ event, data }));
    }
  };

  let socket = null;
  let context = null;
  let capture = null; // MediaStreamAudioSourceNode + worklet, once granted
  let stream = null;
  let closed = false;
  let attempt = 0;
  let retryTimer = null;
  // When the next scheduled buffer should start, on the context clock. Behind
  // `currentTime` means we fell behind and the queue has drained.
  let nextStart = 0;
  let live = []; // scheduled sources, so a barge-in can stop them all
  let playedMs = 0; // of the current utterance, for playback.cancelled
  let analyser = null; // input level, for the panel's meter
  let wantedInput = undefined; // deviceId the streamer picked, if any

  const role = window.bilisamaShell ? "shell" : "browser";
  const url =
    `ws://${location.host}${location.pathname.replace(/\/$/, "")}` + `/audio?role=${role}`;

  const open = () => {
    if (closed) return;
    socket = new WebSocket(url);
    socket.binaryType = "arraybuffer";
    socket.onopen = () => {
      attempt = 0;
      onOwner(role);
      startCapture();
    };
    socket.onmessage = (message) => schedule(new Int16Array(message.data));
    socket.onclose = (event) => {
      stopCapture();
      onOwner(null);
      if (closed) return;
      if (event.code === 4409) {
        // Another client holds the devices — the shell, almost always, with
        // this being a developer's tab. Retrying would be a hot loop against a
        // door that is not going to open.
        return;
      }
      const delay = Math.min(8000, 500 * 2 ** attempt) * (0.7 + Math.random() * 0.6);
      attempt += 1;
      retryTimer = setTimeout(open, delay);
    };
    socket.onerror = () => socket.close();
  };

  async function startCapture() {
    if (capture) return;
    try {
      stream = await navigator.mediaDevices.getUserMedia({
        audio: {
          ...(wantedInput ? { deviceId: { exact: wantedInput } } : {}),
          // The whole reason this file exists. Noise suppression and gain
          // control ride along because the same capture chain runs them and
          // a streaming microphone wants both (plan section 3.2).
          echoCancellation: true,
          noiseSuppression: true,
          autoGainControl: true,
        },
      });
    } catch (err) {
      // Denied, or no device. Say so once and stay silent: the reply audio
      // still plays, and the local sounddevice pair is not coming back while
      // this socket holds the claim.
      console.warn("拿不到麦克风：", err);
      onOwner(role, String(err));
      return;
    }
    context = context ?? new AudioContext();
    await context.audioWorklet.addModule(new URL("./capture-worklet.js", import.meta.url));
    const node = new AudioWorkletNode(context, "bilisama-capture");
    node.port.onmessage = (message) => {
      if (socket && socket.readyState === WebSocket.OPEN) socket.send(message.data);
    };
    const source = context.createMediaStreamSource(stream);
    // Through a silent gain into the destination, which looks pointless and is
    // not: Web Audio renders by pulling backwards from the destination, so a
    // worklet with nothing downstream never has process() called at all. The
    // gain is zero because the alternative — microphone into speakers — is a
    // feedback loop.
    const mute = context.createGain();
    mute.gain.value = 0;
    source.connect(node);
    node.connect(mute);
    mute.connect(context.destination);
    // A parallel tap for the meter. fftSize is the smallest the spec allows:
    // this is a bar, not a spectrogram, and the cheapest read wins.
    analyser = context.createAnalyser();
    analyser.fftSize = 32;
    source.connect(analyser);
    capture = { node, source, mute };
  }

  function stopCapture() {
    capture?.source.disconnect();
    capture?.node.disconnect();
    capture?.mute.disconnect();
    capture?.node.port.close();
    capture = null;
    stream?.getTracks().forEach((track) => track.stop());
    stream = null;
  }

  function schedule(samples) {
    if (!samples.length) return;
    context = context ?? new AudioContext();
    // Autoplay policy parks a context until a gesture. In the shell there is
    // rarely one before she first speaks, so ask every time — resume() on a
    // running context is free.
    if (context.state === "suspended") context.resume().catch(() => {});
    if (!live.length && nextStart <= context.currentTime) {
      // The queue drained, so this buffer opens a new utterance. played_ms is
      // "how much of THIS reply was heard" — carrying it across would make the
      // number grow all session and mistrim the memory it feeds.
      playedMs = 0;
      nextStart = 0;
    }
    const buffer = context.createBuffer(1, samples.length, DOWNLINK_RATE);
    const channel = buffer.getChannelData(0);
    for (let i = 0; i < samples.length; i += 1) channel[i] = samples[i] / 32768;

    const source = context.createBufferSource();
    source.buffer = buffer;
    source.connect(context.destination);
    // A small cushion the first time: starting exactly at currentTime means
    // any scheduling jitter lands in the past and the browser drops the head
    // of the first word.
    const start = Math.max(nextStart, context.currentTime + 0.05);
    source.start(start);
    nextStart = start + buffer.duration;
    live.push(source);
    // Per SEGMENT, both of them. One reply plays as many buffers, and the
    // instant between two of them is not the end of speaking — reading it as
    // one is how the backlog comes back (ledger #41).
    report("playback.started");
    source.onended = () => {
      live = live.filter((other) => other !== source);
      playedMs += buffer.duration * 1000;
      report("playback.ended");
    };
  }

  open();

  return {
    /** Stop everything scheduled and report how much of it was heard. */
    clear() {
      const heard = Math.round(playedMs);
      live.forEach((source) => {
        source.onended = null; // no ended receipts for what nobody heard
        try {
          source.stop();
        } catch {
          // Already finished between the barge-in and this loop.
        }
      });
      live = [];
      nextStart = 0;
      playedMs = 0;
      report("playback.cancelled", { played_ms: heard });
    },
    /** 0..1 input loudness right now, or 0 when there is no microphone. */
    level() {
      if (!analyser) return 0;
      const bins = new Uint8Array(analyser.fftSize);
      analyser.getByteTimeDomainData(bins);
      let peak = 0;
      for (const bin of bins) peak = Math.max(peak, Math.abs(bin - 128) / 128);
      return peak;
    },
    /** Reopen capture on a different microphone. */
    async useInput(deviceId) {
      wantedInput = deviceId || undefined;
      stopCapture();
      await startCapture();
    },
    /** Move playback to a different speaker.
     *
     * Worth knowing before reaching for it: the echo canceller references the
     * output the browser considers current, so sending playback somewhere else
     * can quietly cost the cancellation this whole file exists for. The panel
     * says as much next to the control.
     */
    async useOutput(deviceId) {
      context = context ?? new AudioContext();
      if (typeof context.setSinkId !== "function") return false;
      await context.setSinkId(deviceId || "");
      return true;
    },
    /** A short tone down the real playback path, to confirm the right speaker. */
    async test() {
      context = context ?? new AudioContext();
      if (context.state === "suspended") await context.resume().catch(() => {});
      const osc = context.createOscillator();
      const gain = context.createGain();
      osc.frequency.value = 660;
      // Ramped, not switched: a square edge on a speaker is a click, and a
      // click is exactly what someone testing their audio should not hear.
      gain.gain.setValueAtTime(0, context.currentTime);
      gain.gain.linearRampToValueAtTime(0.18, context.currentTime + 0.02);
      gain.gain.linearRampToValueAtTime(0, context.currentTime + 0.35);
      osc.connect(gain);
      gain.connect(context.destination);
      osc.start();
      osc.stop(context.currentTime + 0.36);
    },
    /** Devices the streamer can pick between, named. */
    async devices() {
      // Labels stay empty until permission is granted, so this is worth
      // nothing before capture starts — the panel calls it after.
      const all = await navigator.mediaDevices.enumerateDevices();
      return all
        .filter((device) => device.kind === "audioinput" || device.kind === "audiooutput")
        .map((device) => ({
          id: device.deviceId,
          kind: device.kind,
          label: device.label || "（未命名设备）",
        }));
    },
    close() {
      closed = true;
      clearTimeout(retryTimer);
      stopCapture();
      socket?.close();
    },
  };
}
